lib/Parse/ParseDecl.cpp

```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
     if (UnderlyingTy.isNull())
       return Status;
   }
-  
-  // Parse a 'where' clause if present. These are not supported, but we will
-  // get better QoI this way.
+
+  TrailingWhereClause *TrailingWhere = nullptr;
+  // Parse a 'where' clause if present.
   if (Tok.is(tok::kw_where)) {
-    GenericParamList *unused = nullptr;
-    auto whereStatus = parseFreestandingGenericWhereClause(
-        unused, WhereClauseKind::AssociatedType);
-    if (whereStatus.shouldStopParsing())
-      return whereStatus;
+    if (Context.isSwiftVersion3()) {
+      // Swift 3 doesn't support where clauses on associated types, so let's
+      // keep exactly the same parsing behaviour as it had.
```

Comment:
I think we want the new feature to still work in Swift 3 mode too. This was the decision for nested generics and generic subscripts. Only incompatible changes should be limited to Swift 4 mode.